### PR TITLE
Add OuputVerbosity and use it in assertions

### PR DESCRIPTION
**Pull request by plannigan**
`OuputVerbosity` is a new class that allows for retrieving a verbosity level for a specific output type. If nothing is specifically configured, the application wide verbose value is used.

The first use of this functionality is `verbosity_assertions` which can be used to make the assertion reports output at the most verbose level while still showing a single character for each test case when progress is being reported.

Addresses #11387

### CODE REVIEW

File: src/_pytest/config/__init__.py
Comment (plannigan) [inline src/_pytest/config/__init__.py:1696]: Since many things will add these type of settings, I thought it would be helpful to have this function to ensure it is dong consistently. I'm not completely in love with this level of indirection. But it is necessary since config settings are added to the `Parser` before the `Config` object exists.

File: testing/test_assertion.py
Comment (plannigan) [inline testing/test_assertion.py:20]: With the move away from `getoption()`, this needed to get a bit more complicated. (In the long term, it would probably be beneficial move the code base to a consistent way to access the verbosity level.)

**Comment by plannigan** [comment]
Since this is my first contribution, I'm not sure the best way to get a review for this PR that has been up for 3 weeks. @nicoddemus & @Zac-HD, you both interacted with the issue. Would one of you be able to review this code?

**Comment by nicoddemus** [comment]
Sorry @plannigan, you are not supposed to do anything else other than opening the PR: we dropped the ball for not reviewing it earlier (I personally have been busy). When that happens a ping (like you did) is appreciated, thank you!

I will try to review this in the next couple of days.

File: doc/en/how-to/output.rst
Comment (nicoddemus) [inline doc/en/how-to/output.rst:279]:
```suggestion
:confval:`verbosity_assertions`: Controls how verbose the assertion output should be when pytest is executed. A value of ``2``
```

File: doc/en/how-to/output.rst
Comment (nicoddemus) [inline doc/en/how-to/output.rst:282]:
```suggestion

(Note: currently this is the only option available, but more might be added in the future).
```

File: src/_pytest/assertion/__init__.py
Comment (nicoddemus) [inline src/_pytest/assertion/__init__.py:51]:
```suggestion
            "Higher levels will provide more detailed explanation when an assertion fails."
```

File: src/_pytest/config/__init__.py
Comment (nicoddemus) [reply src/_pytest/config/__init__.py:1696]: Good idea. 👍

File: src/_pytest/config/__init__.py
Comment (nicoddemus) [inline src/_pytest/config/__init__.py:1678]: I think if one passes an invalid `output_type` at the moment, we will get an error (dur to `self._config.getini`).

We probably want to have this classs accessible to plugins and users, via `config.output_verbosity`, so we need to think about its API carefully as we will need to support this officially moving on.

Some points:

1. In tandem with pytest's direction regarding type safety, I think we should use an `Enum` with the output types currently available, something along the lines of:

   ```python
   class VerbosityType(Enum):
       """..."""     

       Global = "global"
       Assertions = "assertions"
   ```

   Instead of a simple `output_type: str` parameter/value.

2. The class name is `OutputVerbosity` and the public methods are `verbosity` and `verbosity_for`, which read a bit redundant: `confg.output_verbosity.verbosity`, `confg.output_verbosity.get_verbosity_for`. I suggest we use simply `get`, defaulting to the `Global` verbosity:

   ```python
   def get(self, type: VerbosityType = VerbosityType.Global) -> int:
       ...
   ```

   Which reads well I think:

   ```python
   config.output_verbosity.get()
   config.output_verbosity.get(VerbosityType.Assertions)
   ```

3. We need to properly document both `OutputVerbosity` and `VerbosityType` in the reference docs (including some usage examples, but possibly we can write those in the class docstring). We also need to expose `VerbosityType` in the `pytest` namespace.

4. Given `OutputVerbosity` is public, we should make `OutputVerbosity.addini` private, as it is intended for internal use only.


I'm not married to the names above, so suggestions and bikeshedding are welcome.

File: changelog/11387.feature.rst
Comment (nicoddemus) [inline changelog/11387.feature.rst:1]: We should also mention `OutputVerbosity` and link to its docs.

**Review by nicoddemus — CHANGES_REQUESTED** [review]
Great work @plannigan, and sorry again for the delay!

The implementation is good, however I think we might need some more interactions given this introduce a new API, which must be done carefully and with care.

File: src/_pytest/config/__init__.py
Comment (plannigan) [reply src/_pytest/config/__init__.py:1678]: 1. I opted for `str` because I was thinking that plugins might want also register types, like any other config setting. Which would mean that it wouldn't be possible to statically define all of the types. However, reflecting on that, I'm not sure that level of extensibility is necessary. It is likely better for `pytest` to say "here are N buckets that a plugin can utilize" and that way users of multiple plugins that leverage this functionality will see consistent behavior without needing to enable M settings.
2. That is fair. It also works nicely with the enum suggestion.
3. Sure
4. I had it as public for the reasons described in 1, but it makes sense to make it private based on that change.

**Review by bluetech — COMMENTED** [review]
I'm not able to do a full review ATM, but left some quick comments.

File: doc/en/how-to/output.rst
Comment (plannigan) [reply doc/en/how-to/output.rst:294]: My intention is to implement some of the other verbosity types I identified once this PR is merged. A different suggestion was to add a note that only one is currently supported. That seems sufficient to me if "life happens" and it takes me some time to get to adding other types.

File: doc/en/reference/reference.rst
Comment (plannigan) [reply doc/en/reference/reference.rst:1834]: Two reasons:

* My use case is to always enable `-vv` for assertions without increasing the verbosity of the rest of the output. Setting that in the configuration file allows this to be enabled for a project without being explicitly added each time `pytest` is run.
* My initial implementation took an arbitrary string when adding a new level. The thought was that any plugin might add their own type. This would mean that there could potentially be a large number of command line options related to fine-grained verbosity, which seemed harder to manage. With the move to `pytest` statically defining each type, that will keep the N small, so I'm less apposed to the idea.
* With the existence of `addopts` and `--override-ini`, it isn't clear if there are rules about what should be a command line option version a configuration file setting.

File: src/_pytest/config/__init__.py
Comment (plannigan) [reply src/_pytest/config/__init__.py:1675]: I'm open tweaking the name. I generally default to more explicit names.

File: src/_pytest/config/__init__.py
Comment (plannigan) [reply src/_pytest/config/__init__.py:1701]: Like the class name, I'm open tweaking the name. I went with `verbose` to make the command line option name destination (which is how most things access the application wide value now). But `pytest` is already inconsistent:

* command line options name: `--verbosity`
* command line options destination name: `verbose`
* `TerminalReporter` property: `verbosity`

Maybe a follow up cleanup PR would be to make the code base more consistent.

**Comment by plannigan** [comment]
I believe I have addressed the technical comments and documentation requests.

I'm not sure why the doc building step is failing. The error don't seem to be related to the lines I changed. If someone with more experience with `sphinx` can take a look, I would appreciate it.

**Comment by nicoddemus** [comment]
> I'm not sure why the doc building step is failing

The problem was that you used `^^^` for the heading in your new session, while in the rest of the document we used `~~~` for that level of heading. I pushed a fix. 👍

File: doc/en/reference/reference.rst
Comment (nicoddemus) [reply doc/en/reference/reference.rst:1834]: I agree this is something one might want to configure mainly via the configuration file, so a configuration value seems reasonable.

File: src/_pytest/config/__init__.py
Comment (nicoddemus) [reply src/_pytest/config/__init__.py:1701]: I suggest we just remove this option, and let users access it via `.get()`.

Note: I removed and pushed this property in a separate commit, we can revert it if we decide it otherwise.

**Review by nicoddemus — APPROVED** [review]

**Review by bluetech — CHANGES_REQUESTED** [review]
Thanks for the update @plannigan, my comments:

1. The `OutputVerbosity` and `VerbosityType` seem more complicated than necessary to me. What is the advantage over doing it the direct way, i.e. adding the ini directly and maybe adding a getter on `Config`?

2. If we stick with `OutputVerbosity`, let's make it just `Verbosity` -- the `Output` part is entirely redundant IMO.

3. We should have at least one end-to-end test (`pytester.runpytest` test) which shows this working, i.e. changing the output vs. not setting the ini.

**Comment by nicoddemus** [comment]
> i.e. adding the ini directly 

I think this is to handle the fallback to default verbosity when the setting is defined in the ini file.

> and maybe adding a getter on Config?

Can you clarify what you mean here, do you mean adding:

1. `Config.get_verbosity(VerbosityType) -> int` 
2. `Config.get_verbosity(Literal["assertion"]) -> int` 
3. `Config.get_assertion_verbosity() -> int`

**Comment by bluetech** [comment]
> Can you clarify what you mean here, do you mean adding:

Yes, any of these would be fine by me.

**Comment by nicoddemus** [comment]
One nice effect is that it would cut down on the number of things exposed publicly, specially 2) in my suggestion above: we can by default return the default on "unknown" verbosity types, to make the life of plugin authors easier: using a `VerbosityType` enum has many advantages, but one disadvantage is that plugin authors that need to support multiple pytest versions need to do some juggling to see if they have access to `VerbosityType.X`.

Suppose in the future we add `VerbosityType.Rubles` in pytest 9.0, plugins would have to juggle a bit to support pytest 7.x+ up to 9:

```python
if hasattr(config, "output_verbosity") and hasattr(pytest, "VerbosityType") and hasattr(pytest.VerbosityType, "Rubles"):
    rubles_verbosity = config.output_verbosity.get(pytest.VerbosityType, Rubles)
else:
    rubles_verbosity = config.option.verbose
```

While using option 2) above users can go with:

```python
if hasattr(config, "get_verbosity"):
    rubles_verbosity = config.get_verbosity("rubles")
else:
    rubles_verbosity = config.option.verbose
```

**Comment by plannigan** [comment]
The forward comparability argument for pivoting back towards strings makes sense to me. I should have time this weekend to implement that and the absorption of `OutputVerbosity` into `Config`.

**Comment by nicoddemus** [comment]
Thanks @plannigan, appreciate the patience and willing to change the implementation. 👍

File: changelog/11387.feature.rst
Comment (nicoddemus) [inline changelog/11387.feature.rst:5]:
```suggestion
For plugin authors, :attr:`config.get_verbosity <pytest.Config.get_verbosity>` can be used to retrieve the verbosity level for a specific verbosity type.
```

File: testing/test_assertion.py
Comment (nicoddemus) [inline testing/test_assertion.py:75]: We do test all the pieces individually, which is great, but I think we should also include an integration test (using `pytester`) in this file, that runs pytest with different ` verbosity_assertions` levels and checks that the output differs accordingly.

Comment (plannigan) [reply testing/test_assertion.py:75]: Thanks for reminding me to do this. It made me notice that I had missed `saferepr`'s max size.

**Comment by plannigan** [comment]
Looks like one of the jobs failed with a file system error in an area I didn't touch. I don't have permissions to retry the job.

**Comment by nicoddemus** [comment]
> Looks like one of the jobs failed with a file system error in an area I didn't touch. I don't have permissions to retry the job.

Started it again. 👍

File: src/_pytest/config/__init__.py
Comment (bluetech) [inline src/_pytest/config/__init__.py:1692]: The name `_ini_name` is now too general, we should change it to `_verbosity_ini_name` or (my preference) inline it.

File: src/_pytest/config/__init__.py
Comment (bluetech) [inline src/_pytest/config/__init__.py:1659]: The docstring should explain what `None`/the default means.

File: src/_pytest/config/__init__.py
Comment (bluetech) [inline src/_pytest/config/__init__.py:1683]: Let's add a unit test which covers this branch.

File: src/_pytest/config/__init__.py
Comment (bluetech) [inline src/_pytest/config/__init__.py:1697]: Same here, the name `_add_ini` is too general now.

Also, IMO this function doesn't add much value, I think it would be better to reduce the indirection and inline it. But since it's private then it's OK if you want to keep it.

File: src/_pytest/config/__init__.py
Comment (bluetech) [inline src/_pytest/config/__init__.py:1687]: The `auto` value should be documented.

Also, why did you choose "auto"?

File: testing/test_assertion.py
Comment (bluetech) [inline testing/test_assertion.py:1910]: Write the real name here; as an integration test it should tell us if the actual thing the user writes regresses.

File: testing/test_assertion.py
Comment (bluetech) [inline testing/test_assertion.py:1913]:
```suggestion
    result = pytester.runpytest(p)
```

**Review by bluetech — CHANGES_REQUESTED** [review]
Haven't had time for a full review again, but left some comments, please take a look.

File: src/_pytest/config/__init__.py
Comment (plannigan) [reply src/_pytest/config/__init__.py:1700]: Added some documentation. Also, it was the value used in [this related PR](https://github.com/pytest-dev/pytest/pull/10588/files#diff-0e2d32c88d3fa7edb466f8344dae556d52632e608e7414371f0537ce9dcc3a88R31) on the same topic. I'm open to changing it if there is a better suggestion.

File: src/_pytest/config/__init__.py
Comment (plannigan) [reply src/_pytest/config/__init__.py]: I prefer to keep it because it will avoid duplication as more types are added. It also ensures that the same default is used across verbosity types. If someone adds an Nth type, but uses "global", then `get_verbosity()` will raise an exception.

File: testing/test_assertion.py
Comment (plannigan) [reply testing/test_assertion.py]: Addressed, but in my defense, I copied this from other examples in the file :smile:.

File: doc/en/reference/reference.rst
Comment (bluetech) [inline doc/en/reference/reference.rst:1003]: Better to keep it

File: src/_pytest/assertion/truncate.py
Comment (bluetech) [inline src/_pytest/assertion/truncate.py:12]: Better to keep it

File: src/_pytest/config/__init__.py
Comment (bluetech) [inline src/_pytest/config/__init__.py:1694]: I think it would be better to avoid this check, so that plugins can add their own verbosity types. We would instead check if the ini exists.

File: src/_pytest/config/__init__.py
Comment (bluetech) [inline src/_pytest/config/__init__.py:1657]:
```suggestion
    _VERBOSITY_INI_DEFAULT: Final = "auto"
```

**Review by bluetech — CHANGES_REQUESTED** [review]
Thanks for the update @plannigan, please see my comment on `_KNOWN_VERBOSITY_TYPES`. Other than that it looks good!

File: src/_pytest/config/__init__.py
Comment (plannigan) [reply src/_pytest/config/__init__.py:1694]: I like the idea of checking the ini data instead (one less thing to manually manage). 

In a [previous conversation](https://github.com/pytest-dev/pytest/pull/11473#discussion_r1369828592), there was a conversation about plugins creating their own levels. I think I'm still in the "pytest can enumerate a few buckets" camp. @nicoddemus Do you have any thoughts? If we did allow plugins to create types, `_add_verbosity_ini` would have to be made public and some more documentation updates.

Comment (nicoddemus) [reply src/_pytest/config/__init__.py:1694]: I think we can keep it simple for now, making this internal only; if the need comes later, we can discuss how to make this public -- better to err on the side of caution, as it is easy to make something public later, rather than making something public and then regretting it later.

**Review by bluetech — APPROVED** [review]
LGTM! I know this feature will make a lot of people happy :)

I'll let @nicoddemus do the final review/merge.

**Review by nicoddemus — APPROVED** [review]

**Comment by nicoddemus** [comment]
Thanks everyone!

**Pull request merged**
